device/core/src/device_method/method_client.ts

```diff
@@ -0,0 +1,115 @@
+        /*Codes_SRS_NODE_DEVICE_METHOD_CLIENT_16_009: [The `invokeMethod` method shall call the `setOptions` method on the `RestApiClient` with its options as argument to make sure the CA certificate is populated.]*/
+        this._restApiClient.setOptions(this._options);
+        /*Codes_SRS_NODE_DEVICE_METHOD_CLIENT_16_010: [The `invokeMethod` method shall construct the HTTP request path as `/twins/encodeUriComponentStrict(<targetDeviceId>)/methods` if the target is a device.]*/
+        let path = `/twins/${encodeUriComponentStrict(deviceId)}`;
+        /*Codes_SRS_NODE_DEVICE_METHOD_CLIENT_16_011: [The `invokeMethod` method shall construct the HTTP request path as `/twins/encodeUriComponentStrict(<targetDeviceId>)/modules/encodeUriComponentStrict(<targetModuleId>)/methods` if the target is a module.]*/
+        if (moduleId) {
+          path += `/modules/${encodeUriComponentStrict(moduleId)}`;
+        }
+        path += '/methods';
+
+        /*Codes_SRS_NODE_DEVICE_METHOD_CLIENT_16_012: [The `invokeMethod` method shall call `RestApiClient.executeApiCall` with:
+          - `POST` for the HTTP method argument.
+          - `path` as defined in `SRS_NODE_DEVICE_METHOD_CLIENT_16_010` and `SRS_NODE_DEVICE_METHOD_CLIENT_16_011`
+          - 2 custom headers:
+            - `Content-Type` shall be set to `application/json`
+            - `x-ms-edge-moduleId` shall be set to `<deviceId>/<moduleId>` with `deviceId` and `moduleId` being the identifiers for the current module (as opposed to the target module)
+          - the stringified version of the `MethodParams` object as the body of the request
+          - a timeout value in milliseconds that is the sum of the `connectTimeoutInSeconds` and `responseTimeoutInSeconds` parameters of the `MethodParams` object.]*/
+        const body = JSON.stringify(methodParams);
+        const methodTimeout = 1000 * (methodParams.connectTimeoutInSeconds + methodParams.responseTimeoutInSeconds);
+        this._restApiClient.executeApiCall('POST', path, this._httpHeaders, body, methodTimeout, (err, result) => {
+          if (err) {
+            /*Codes_SRS_NODE_DEVICE_METHOD_CLIENT_16_013: [The `invokeMethod` method shall call its callback with an error if `RestApiClient.executeApiCall` fails.]*/
+            callback(err);
+          } else {
+            /*Codes_SRS_NODE_DEVICE_METHOD_CLIENT_16_014: [The `invokeMethod` method shall call its callback with the result object if the call to `RestApiClient.executeApiCall` succeeds.]*/
+            callback(null, result as MethodResult);
+          }
+        });
+      }
+    });
+  }
+
+  setOptions(options: any): void {
+    /*Codes_SRS_NODE_DEVICE_METHOD_CLIENT_16_001: [The `setOptions` method shall merge the options passed in argument with the existing set of options used by the `MethodClient`.]*/
+    this._options = _.merge(this._options, options);
+  }
+
+  updateSharedAccessSignature(sharedAccessSignature: string, callback: (err?: Error) => void): void {
```

Comment:
You're right... I had that as I was thinking of an old historical code path - but since we have AuthenticationProviders, this problem is solved elegantly and I actually don't need it.
(user calls ModuleClient.updateSharedAccessSignature - which in turn will call the transport updateSharedAccessSignature method, which will update the AuthenticationProvider. since getDeviceCredentials is called before any API call, it'll always be up to date and I don't need this function at all. brilliant :) thanks for pointing it out!

---
In reply to: [197007121](https://github.com/Azure/azure-iot-sdk-node/pull/322#discussion_r197007121) [](ancestors = 197007121)